File: plugins/InspectionGadgets/test/com/siyeh/igtest/dependency/suspicious_package_private_access/src/AccessingProtectedMembers.java
package xxx;

class AccessingProtectedMembersNotFromSubclass {
  void foo() {
    ProtectedMembers aClass = new ProtectedMembers();
    aClass.<warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>();
    ProtectedMembers.<warning descr="Method ProtectedMembers.staticMethod() is protected and used not through a subclass here, but declared in a different module 'dep'">staticMethod</warning>();
    new <warning descr="Constructor ProtectedConstructors.ProtectedConstructors() is protected and used not through a subclass here, but declared in a different module 'dep'">ProtectedConstructors</warning>();
    new <warning descr="Constructor ProtectedConstructors.ProtectedConstructors(int) is protected and used not through a subclass here, but declared in a different module 'dep'">ProtectedConstructors</warning>(1);
    new ProtectedConstructors() {};
    new ProtectedConstructors(1) {};
  }

  void baz() {
    class LocalSubclass extends ProtectedMembers {
      void bar() {
        method();
        staticMethod();
      }
    }
  }
}

class AccessingProtectedMembersFromSubclass extends ProtectedMembers {
  void foo() {
    method();
    staticMethod();
    ProtectedMembers.staticMethod();

    ProtectedMembers aClass = new ProtectedMembers();
    aClass.<warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>();
    AccessingProtectedMembersFromSubclass myInstance = new AccessingProtectedMembersFromSubclass();
    myInstance.method();

    ProtectedMembers.StaticInner inner1;
    StaticInner inner2;

    new Runnable() {
      public void run() {
        method();
        staticMethod();
      }
    };

    class LocalClass {
      void baz() {
        method();
        staticMethod();
      }
    }
  }

  public static class StaticInnerImpl1 extends ProtectedMembers.StaticInner {
  }

  public static class StaticInnerImpl2 extends StaticInner {
  }

  public class OwnInner {
    void bar() {
      method();
      staticMethod();
    }
  }

  public static class OwnStaticInner {
    void bar() {
      staticMethod();
    }
  }
}

class AccessingDefaultProtectedConstructorFromSubclass extends ProtectedConstructors {
}

class AccessingProtectedConstructorFromSubclass extends ProtectedConstructors {
  AccessingProtectedConstructorFromSubclass() {
    super(1);
  }
}

File: jvm/jvm-analysis-kotlin-tests/testData/codeInspection/suspiciousPackagePrivateAccess/src/AccessingProtectedMembers.kt
package xxx

class AccessingProtectedMembersNotFromSubclass {
  fun foo() {
    val aClass: ProtectedMembers = ProtectedMembers()
    aClass.<warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>()
    ProtectedMembers.<warning descr="Method ProtectedMembers.staticMethod() is protected and used not through a subclass here, but declared in a different module 'dep'">staticMethod</warning>()
    <warning descr="Constructor ProtectedConstructors.ProtectedConstructors() is protected and used not through a subclass here, but declared in a different module 'dep'">ProtectedConstructors</warning>()
    <warning descr="Constructor ProtectedConstructors.ProtectedConstructors(int) is protected and used not through a subclass here, but declared in a different module 'dep'">ProtectedConstructors</warning>(1)
  }
}

@Suppress("UNUSED_VARIABLE")
class AccessingProtectedMembersFromSubclass : ProtectedMembers() {
  fun foo() {
    method()
    staticMethod()
    ProtectedMembers.staticMethod()

    val aClass = ProtectedMembers()
    aClass.<warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>()
    val myInstance = AccessingProtectedMembersFromSubclass()
    myInstance.method()

    var inner1: ProtectedMembers.StaticInner
    var inner2: StaticInner

    val runnable = object : Runnable {
      override fun run() {
        <warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>()
        <warning descr="Method ProtectedMembers.staticMethod() is protected and used not through a subclass here, but declared in a different module 'dep'">staticMethod</warning>()
      }
    }
  }

  private val runnable = object : Runnable {
    override fun run() {
      <warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>()
      <warning descr="Method ProtectedMembers.staticMethod() is protected and used not through a subclass here, but declared in a different module 'dep'">staticMethod</warning>()
    }
  }

  private val obj = object : StaticInner() {}

  private class StaticInnerImpl1 : ProtectedMembers.StaticInner()

  private class StaticInnerImpl2 : StaticInner()

  private inner class OwnInnerClass {
    fun bar() {
      <warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>()
      <warning descr="Method ProtectedMembers.staticMethod() is protected and used not through a subclass here, but declared in a different module 'dep'">staticMethod</warning>()
    }
  }
}

class AccessingDefaultProtectedConstructorFromSubclass : ProtectedConstructors()

class AccessingProtectedConstructorFromSubclass : ProtectedConstructors(1)

val objectAccessingDefaultProtectedConstructorFromSubclass = object : ProtectedConstructors() {}

val objectAccessingProtectedConstructorFromSubclass = object : ProtectedConstructors(1) {}

class AccessingProtectedMembersFromObjectLiteral {
  fun bar1() {
    object : ProtectedMembers() {
      fun bar2() {
        method()
        object : Runnable {
          override fun run() {
            <warning descr="Method ProtectedMembers.method() is protected and used not through a subclass here, but declared in a different module 'dep'">method</warning>()
          }
        }
      }
    }
  }
}

//KT-35296: Must not produce false positive warnings for package-private empty constructor.
class AccessProtectedSuperConstructor : PackagePrivateEmptyConstructor {
  constructor(i: Int) : super(i)

  constructor(i: Int, i2: Int): this(i + i2)
}

class AccessProtectedSuperConstructorWithArgument(i: Int) : PackagePrivateEmptyConstructor(i) {
  constructor(i: Int, i2: Int): this(i + i2)
}